### ISSUE TITLE
optimize loops for the common cases

### ISSUE THREAD

**Issue by rurban** [perl11/cperl#138]
add special faster ops for pp_iter:
- iter_ary `for (@array)`
- iter_lazyiv `for (0..9)`
  without REVERSED support, and a new compile-time check for invalid decrementing range.

---
- enteriter is always followed by iter, and: **enteriter - iter - and**
- iter is either after enteriter (first iter) or unstack (subsequent): **unstack - iter - and**
  unstack being the nextop.
- `for $sv ()` is always converted to a rv2gv - gv tree, implicit `for ()` uses a single gv only.
A non-lexical scalar itervar $sv can be converted to a single gvsv without OPf_REF, to seperate it from `for \$sv ()`. enteriter/cx_pushloop_for wants a gv/sv CXp_FOR_GV. Replace CXp_FOR_GV with CXp_FOR_SV. Does a GVSV -> GvSV(defgv) make gv - defgv slower?
This should be changed to allow fast non-lexical GVSV itervars, initializing the new glob with rv2gv only once, not at every loop step. Only implicit $_ is pretty fast this way, as it does not need a rv2gv deref.

inline OP_AND into OP_ITER. The AND always acts either on yes or no (produced by iter). so convert iter from base to an UNOP or LOGOP, with first (ignored), other: yes (continue), next: no (leaveloop).

---

old analysis for 5.22-style loops:
case 1: `for (...) {}`

=> **pp_iter_def** op

  stacked
  CXt_LOOP_FOR
  itervar: GvSV(PL_defgv)  
  av_is_stack (!cx->blk_loop.state_u.ary.ary)
  !(PL_op->op_private & OPpITER_REVERSED)
  !(SvMAGICAL(av) || AvREIFY(av))
  !(cx->cx_type & CXp_FOR_LVREF)

case 2: `for my $var (...) { }`

=> **pp_iter_lex** op

  !(PL_op->op_private & OPpITER_REVERSED)
  !(SvMAGICAL(av) || AvREIFY(av))
  !(cx->cx_type & CXp_FOR_LVREF)

case 3: `for (0..10) {}`

with the block consisting of const iv ranges. the temp. setiv can be optimized then (to defgv or a lexical)

**Comment by rurban** [comment]
Will do the iter-and merge later.
released in 5.25.0c

**Comment by rurban** [comment]
result with `Porting/bench.pl --tests=/^loop/`: 13% faster overall, oops 2x faster (skipped another runtime  NULL).
```
          new     old
       ------ -------
    Ir 100.00   49.15
    Dr 100.00  119.73
    Dw 100.00   85.16
  COND 100.00   84.55
   IND 100.00   99.06
```
